### PR TITLE
builder: fix layer lifecycle leak

### LINKED ISSUE #135: Multi-stage cache bringing back deleted files
* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

### Expected behavior

I have a simple Dockerfile with two stages. First I copy some files, run a small build script and then delete the original files. My second stage copies the result from the first stage. I would expect the deleted files to not be there, but sometimes they show up unless I use `--no-cache` when building.

### Actual behavior

Deleted files are somehow left in the container.

### Steps to reproduce the behavior

To reproduce create an empty folder with this `Dockerfile`:
```
FROM python:3.6 AS compiled

RUN mkdir -p /usr/src/app
WORKDIR /usr/src/app

COPY . /usr/src/app

RUN python -m compileall -b .
RUN find . -name "*.py" -type f | xargs rm -fv

# actual image
FROM python:3.6

RUN mkdir -p /usr/src/app
WORKDIR /usr/src/app

COPY --from=compiled /usr/src/app .

ENTRYPOINT ls
```

Then run this command:

```
touch a.py && rm -f b.py && docker build --no-cache -t test . && docker run --rm test && cp a.py b.py && docker build -t test . && docker run --rm test
```

You'll see the first time the container runs (middle of the output), you only get `a.pyc` as the result of `ls`. But the second time it runs, the removed `.py` files come back to life.

```
[....]
removed ‘./a.py’
[....]
Dockerfile
a.pyc
[....]
 ---> Using cache
[....]
 ---> Running in 0178c7c81be0
removed ‘./b.py’
removed ‘./a.py’
[....]
Dockerfile
a.py
a.pyc
b.py
b.pyc
```

My current workaround for this is to always disable the cache, but my builds become much slower. I've also noticed, as the example shows, that the input of the first stage must be different for this to happen (it doesn't happen without `cp a.py b.py`).

**Output of `docker version`:**

```
Docker version 17.09.0-ce, build afdb6d4
```

**Output of `docker info`:**

```
Containers: 74
 Running: 0
 Paused: 0
 Stopped: 74
Images: 487
Server Version: 17.09.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 502
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.10.0-37-generic
Operating System: Ubuntu 17.04
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 990MiB
Name: ubuntu
ID: TAF3:JIHX:L4NV:KJ5A:AYPX:J4IU:LB66:KGJP:ZUAK:UVEQ:TLAJ:OKFE
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.)**

I'm running on a fully up-to-date Ubuntu 17.04 running inside VirtualBox 5.1.30.

### LINKED ISSUE #35641: Files leaking to other images in a multi-stage build
**Description**

Files can leak from one image to a different one in a multi-stage build process. I was able to isolate the bug to the following minimalist case.

**Steps to reproduce the issue:**
1. `echo 'This file should only appear in the first image.' >afile`
2. `echo 'This file should only appear in the second image.' >bfile`
3. Save this `Dockerfile`:
```
FROM alpine
WORKDIR /root
COPY afile .

FROM alpine
WORKDIR /root
COPY bfile .
```
4. `docker build .`
5. `docker run --rm <second image> ls`

**Describe the results you received:**

```
afile
bfile
```

**Describe the results you expected:**

```
bfile
```

**Additional information you deem important (e.g. issue happens only occasionally):**
The bug is consistent on this machine but is not present on a different virtual server with CentOS 7, XFS, and devicemapper.

As can be seen from the log, the `WORKDIR` instruction is cached:
```
Sending build context to Docker daemon  4.096kB
Step 1/6 : FROM alpine
 ---> 665ffb03bfae
Step 2/6 : WORKDIR /root
 ---> 24e2272aca94
Removing intermediate container ebdaacce5267
Step 3/6 : COPY afile .
 ---> 9507351e2ff4
Step 4/6 : FROM alpine
 ---> 665ffb03bfae
Step 5/6 : WORKDIR /root
 ---> Using cache
 ---> 24e2272aca94
Step 6/6 : COPY bfile .
 ---> b7a5fa8bdc29
Successfully built b7a5fa8bdc29
```

`ls` lists no files on the common intermediate layer (24e2272aca94).

If I add another unique no-op instruction before the `COPY` to break the caching at that layer:
```
FROM alpine
WORKDIR /root
RUN echo no-op layer 1
COPY afile .

FROM alpine
WORKDIR /root
RUN echo no-op layer 2
COPY bfile .
```
, the files are where they should be. I will be using this approach as a work-around for now.

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:18 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 11
 Running: 11
 Paused: 0
 Stopped: 0
Images: 190
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.8.0-58-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 25.51GiB
Name: ubuntu1
ID: ZSC7:3VAV:U3JD:XIUN:OLEV:HS2T:GTYX:RE6E:WQHC:4CDB:WGWI:IQKF
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
This on a physical server. `/etc/docker/daemon.json` only contains:
```
{
"storage-driver": "overlay2"
}
```

### PR DESCRIPTION
**Pull request by tonistiigi**
fixes https://github.com/docker/for-linux/issues/135
fixes #35641

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

### CODE REVIEW

File: builder/dockerfile/internals.go [hunk 1]
```diff
@@ -114,8 +115,8 @@ func (b *Builder) commitContainer(dispatchState *dispatchState, id string, conta
 	return err
 }
 
-func (b *Builder) exportImage(state *dispatchState, imageMount *imageMount, runConfig *container.Config) error {
-	newLayer, err := imageMount.Layer().Commit()
+func (b *Builder) exportImage(state *dispatchState, layer builder.RWLayer, parent *imageMount, runConfig *container.Config) error {
```
Comment (dnephin) [inline builder/dockerfile/internals.go]: only `parent.Image()` is used, so maybe accept `parent *image.Image` instead of `imageMount`?

**Review by dnephin — APPROVED** [review]
LGTM

File: builder/dockerfile/internals.go (hunk 1 above)
Comment (tonistiigi) [reply builder/dockerfile/internals.go]: replaced with `builder.Image`

**Review by dnephin — APPROVED** [review]

**Comment by tonistiigi** [comment]
cc @vdemeester @AkihiroSuda

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Pull request merged**